Recipe 검색 기능 고도화에 Elastic Search 사용

BuchuKim opened the pull request:
## 개요
멘토링 시간에 말씀드렸던 일들을 진행했습니닷. 현재 RecipeDocument라는 네이밍(?)으로 index에 저장될 document 객체를 만들었고, 조회가 필요할 때 한꺼번에 불러와져야할 tag, ingredient, step을 nested field로 두었어요..
1. 특정 tagName이 전부 포함된 레시피를 검색할 수 있고
2. 제목, 걸리는 시간, 난이도, serving(인분?) 각각의 조건에 따라 검색할 수 있는
endpoint를 추가했어요!!!!!

## 작업사항
- 이거 했습니다 🥲 근데 일단 Recipe만 index로 존재하는 상태라 토픽이 하나인데, 인덱스로 저장할 엔티티 종류가 많아지면 토픽도 늘려야할까요?
<img width="900" alt="image" src="https://github.com/f-lab-edu/recipetory/assets/69233405/a9989dd7-b2a8-4f76-9816-0295e4632229">

- 생각보다 변경 파일이 많은데, 조회 전용 객체를 Recipe -> RecipeDocument로 바꾸면서 기존에 Recipe 자체를 이용하던 서비스 역시 코드가 수정되었기 때문입니당!
- `RecipeSearchController`, `RecipeSearchService`를 분리했고,,
- `ElasticSearchClient`를 이용한 쿼리는 `RecipeDocumentQueryService`라는 놈에 있습니다 (얘에 능숙해지느라 좀.. 시간이 걸림.. ㅈㅅㅈㅅㅈㅅ)

## ??
- es는 transactional이 적용 안되므로.. entity update가 있는 곳에 역시 save()를 해줘야하는거겠죠? create 할 때와 같은 과정을 거치면 되는걸까요..

BuchuKim committed:
chore : elastic search 기본 config

BuchuKim committed:
chore : document kafka config
- Recipe가 생성될 때, Document 생성 작업은 생성 message를 kafka에 넘깁니다

BuchuKim committed:
feat (#30) : `@Document` 엔티티 및 nested field 구성

BuchuKim committed:
feat (#30) : recipe search에 Elastic Search 사용
- Search용 Controller, Service 분리
- RecipeDocument를 조회 전용 객체로 구성
- RecipeDocumentQueryService에 esClient를 이용한 es query 작성

BuchuKim committed:
test : testing document query
- RecipeRepository 구현체 수정

BuchuKim committed:
feat (#30) : 기존 jpa repository에 JPQL 추가
- es와 조회 속도 비교를 위함입니다

BuchuKim committed:
fix : query parameter 수정

f-lab-fi commented on src/main/java/com/recipetory/recipe/infrastructure/RecipeJpaRepository.java:
```diff
@@ -10,4 +18,34 @@ public interface RecipeJpaRepository extends JpaRepository<Recipe,Long> {
+
+    List<Recipe> findByAuthor(User author);
+
+    /**
+     * tagNames의 tagName을 전부 포함하는 레시피들을 검색한다.
+     * @param tagNames
+     * @param tagSize
+     * @return
+     */
+    @Query("SELECT DISTINCT r FROM Recipe r " +
+            "JOIN FETCH r.author WHERE " +
+            ":tagSize > 0 AND :tagSize = " +
+            "(SELECT COUNT(t) FROM Tag t WHERE " +
+            "t.recipe = r AND t.tagName IN :tagNames)")
+    List<Recipe> findByTags(
+            @Param("tagNames") List<TagName> tagNames,
+            @Param("tagSize") int tagSize,
+            Pageable pageable);
+
+    @Query("SELECT DISTINCT r FROM Recipe r " +
+            "JOIN FETCH r.author " +
+            "WHERE r.title LIKE %:title% " +
+            "AND (r.recipeInfo.cookingTime = :cookingTime OR :cookingTime = 'UNDEFINED')" +
+            "AND (r.recipeInfo.difficulty = :difficulty OR :difficulty = 'UNDEFINED')" +
+            "AND (r.recipeInfo.serving = :serving OR :serving = 'UNDEFINED')")
+    List<Recipe> findByRecipeInfo(
+            @Param("title") String title,
+            @Param("cookingTime") CookingTime cookingTime,
+            @Param("difficulty") Difficulty difficulty,
+            @Param("serving") Serving serving);
```
이게 멘토링 시간에 봤던 그녀석이군요

BuchuKim replied:
사실 얘를 테스트목적 이외에 쓸 일은 없겠지만 아까워서 버릴 수 없었어요....🥹

f-lab-fi commented on src/main/java/com/recipetory/recipe/infrastructure/RecipeDocumentQueryService.java:
```diff
@@ -0,0 +1,104 @@
+            infoQueries.add(MatchQuery.of(q ->
+                    q.field("title").query(title))._toQuery());
+        }
+        if (cookingTime != CookingTime.UNDEFINED) {
+            System.out.println("undefined가 아니빈다!");
+            infoQueries.add(NestedQuery.of(
```
?

BuchuKim replied:
부끄럽습니다....;;;;;;;;; 😳🙈;;;

f-lab-fi commented on src/main/java/com/recipetory/recipe/infrastructure/RecipeDocumentQueryService.java:
```diff
@@ -0,0 +1,104 @@
+     * @param serving
+     * @return
+     */
+    public List<RecipeDocument> findByRecipeInfo(
+            String title, CookingTime cookingTime, Difficulty difficulty, Serving serving
+    ) {
+        // title이 빈칸이거나, recipeInfo 항목 타입이 UNDEFIEND이면 검색 조건에서 제외
+        List<Query> infoQueries = new ArrayList<>();
+        if (title != "") {
+            infoQueries.add(MatchQuery.of(q ->
+                    q.field("title").query(title))._toQuery());
+        }
+        if (cookingTime != CookingTime.UNDEFINED) {
+            System.out.println("undefined가 아니빈다!");
+            infoQueries.add(NestedQuery.of(
+                    n -> n.path("recipeInfo").query(
+                            q -> q.match(
+                                    s -> s.field("recipeInfo.cookingTime")
+                                            .query(cookingTime.name()))))._toQuery());
+        }
+        if (difficulty != Difficulty.UNDEFINED) {
+            infoQueries.add(NestedQuery.of(n ->
+                    n.path("recipeInfo").query(
+                            q -> q.match(
+                                    s -> s.field("recipeInfo.difficulty")
+                                            .query(difficulty.name()))))._toQuery());
+        }
+        if (serving != Serving.UNDEFINED) {
+            infoQueries.add(NestedQuery.of(n ->
+                    n.path("recipeInfo").query(
+                            q -> q.match(
+                                    s -> s.field("recipeInfo.serving")
+                                            .query(serving.name()))))._toQuery());
+        }
```
이 부분은 개선을 좀 해보고 싶은데, 방법이 없을까요 ㅋ..
(조건이 늘어나면 if문이 계속 추가되는 형태인데..)

BuchuKim replied:
연속된 if문 3개이상 심하게 보기 안좋죠,, 근데 조건이 여기서 더 늘어나긴 힘들 것 같다는 어느정도의 판단 하였는데.. 어떻게 개선을 할 수 있을까요?
`CookingTime`, `Difficulty`, `Serving` 모두 각각의 Enum이라.. 아예 enum에서 쿼리를 짜게 한다.?는 domain -> infra/application에 의존이 생기는 셈이고
if문 지우고 쿼리 자체에 "항목이 `UNDEFINED`일 경우 match 쿼리를 추가하지 않는다" 로 구성하기엔 머리아프고..

f-lab-fi commented on src/main/java/com/recipetory/recipe/application/RecipeSearchService.java:
```diff
@@ -0,0 +1,80 @@
+     */
+    @Transactional(readOnly = true)
+    public RecipeListDto findRecipeByRecipeInfo(
+            String title, String cookingTime, String difficulty, String serving) {
+        List<RecipeDocument> found = recipeRepository.findByRecipeInfo(
+                title,
+                CookingTime.fromString(cookingTime),
+                Difficulty.fromString(difficulty),
+                Serving.fromString(serving));
```
프론트엔드랑 백엔드가 주고 받는 값이 '10분 이내' 같은 값이면 좀 어색할 수도 있을 것 같아요.
만약 이런식으로 바뀐다면 어떨까요?

사용자들이 '10분 이내'라는 표현보다 '10분 미만'이라는 표현을 더 선호한다고 밝혀져서 UI 상에서 '10분 미만'으로 표기를 바꿨다면, 프론트엔드와 백엔드가 동시에 어떻게 바뀌어야할지?
그리고 이것이 단순히 사용자들이 화면의 화면에 어떻게 보이느냐에 따라 바뀌어야한다는게 어색한 포인트 같습니다.

BuchuKim replied:
현재 Enum class의 static 메소드인 `fromString(String input)`이 받고있는건 Enum value name(= 클래스 이름) 자체인데요.. 반환하는 값은 아래와 같습니다.
```java
Arrays.stream(CookingTime.values())
        .filter(cookingTime -> cookingTime.name() // enum 클래스 이름을 돌며
                .equalsIgnoreCase(input)) // input string과 같은 이름을 찾는다
        .findAny().orElse(UNDEFINED); // 없으면 UNDEFINED
```
"10분 미만"이라는 한글 자체보다 `LESS_THAN_10`이라는 enum name url 쿼리 파라미터로 들어간 상황입니다.
멘토님 말씀하신 상황이 와도 프론트에서는 UI 버튼의 문구만 수정하고 url 빌딩을 할 땐 그대로 `LESS_THAN_10`을 사용해도 될텐데.. enum name의 의미 자체가 사라지지 않는다면 괜찮지 않을까요? 응답에서 enum의 필드값을 직접 사용하는 상황이라면 그 값인 description을 바꾸면 될듯.. 합니다..
<img width="307" alt="image" src="https://github.com/f-lab-edu/recipetory/assets/69233405/36f54064-b4a6-4833-91a2-19c19a241790">

f-lab-fi commented on src/main/java/com/recipetory/recipe/infrastructure/RecipeRepositoryImpl.java:
```diff
@@ -30,12 +30,31 @@ public Optional<Recipe> findById(Long id) {
+        return recipeQueryService.findByTagNames(tagNames);
     }
 
     @Override
     public void deleteById(Long id) {
         recipeJpaRepository.deleteById(id);
+        recipeDocumentRepository.deleteById(id);
     }
```
얘는 왜 이벤트 발행해서 처리 안해주나요 ㅎ

BuchuKim replied:
ㅜㅜ 이번에 다른데에 집중하다보니 ~일단 구성해두고 나중에 수정하자~ 식으로 대강 넘어간게 많네유

f-lab-fi commented on src/test/java/com/recipetory/recipe/infrastructure/RecipeEsTest.java:
```diff
@@ -0,0 +1,111 @@
+    public void setUp() {
+        recipeDocumentRepository.deleteAll();
+    }
+    @Test
+    @DisplayName("tag가 모두 들어있는 레시피를 검색한다.")
+    public void tagsSearchTest() {
+        // given : recipe1은 BOIL, DIET, BREAD 태그를
+        //         recipe2는 BOIL, DIET 태그를 갖는다.
```
테스트 이름이 `tag가 모두 들어있는 레시피를 검색한다.` 가 되면 안되지 않나요?!
지금은 태그가 3개더라도 나중에 4개가 됐을 때도 지금 이 테스트는 수정되지 않아도 될텐데 그때에는 테스트 이름이랑 안맞는게 아닌지?

BuchuKim replied:
(무슨 뜻일까요..) O o .
그럼 테스트 이름이 어떻게 되어야하죠? 실제로 태그 이름이 몇 개가 되든(enum 범위 전체가 되어도 상관없음) 테스트는 수정되지 않아도 됩니다만 ..

BuchuKim replied:
멘토님의 추가 답변)
> Tag가 4개가 되면 그때에도 ‘tag가 모두 들어있는’ 이라고 표현할 수 있을까요?
> 결과적으로 ‘검색하려는 tag가 모두 들어있는 레시피를 검색한다.’가 더 정확하지 않을까 생각했습니다.

'tag가 모두 들어있다'의 의미가, "존재하는" tag 전체인지 / "검색하려는" tag 전체인지 모호할 수 있다는 말씀 동감하며 테스트 이름 수정하겠습니닷

BuchuKim committed:
feat : recipe document delete시 event 처리

BuchuKim committed:
test : kafka test 범위 변경
- ~ 메세지 발행, 메세지 listen ~

BuchuKim committed:
fix : 메소드 이름, 주석
- `NotiicationListener` 메소드 이름을 조금 더 명확하게 수정
- `DocumentListener`의 삭제 메소드에 설명 주석 추가

BuchuKim committed:
fix (#30) : es 쿼리 로직 수정
- 항목에 따라 if문을 늘리는 대신 `RecipeQueryHelper` 클래스 이용
